logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/user/ObserveUserInfoUseCase.kt

```diff
@@ -0,0 +1,83 @@
+package com.wire.kalium.logic.feature.user
+
+import com.wire.kalium.logic.CoreFailure
+import com.wire.kalium.logic.StorageFailure
+import com.wire.kalium.logic.data.team.Team
+import com.wire.kalium.logic.data.team.TeamRepository
+import com.wire.kalium.logic.data.user.OtherUser
+import com.wire.kalium.logic.data.user.UserId
+import com.wire.kalium.logic.data.user.UserRepository
+import com.wire.kalium.logic.data.user.type.UserType
+import com.wire.kalium.logic.functional.Either
+import com.wire.kalium.logic.functional.fold
+import com.wire.kalium.logic.functional.foldEither
+import com.wire.kalium.logic.functional.mapLeft
+import com.wire.kalium.logic.wrapStorageRequest
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.filter
+import kotlinx.coroutines.flow.firstOrNull
+
+/**
+ * Use case that allows observing the user details of a user locally,
+ * or request it from API and save to DB, if there is no local data for such user.
+ */
+fun interface ObserveUserInfoUseCase {
+    /**
+     * Use case [GetUserInfoUseCase] operation
+     *
+     * @param userId the target user identifier
+     * @return a [GetUserInfoResult] indicating the operation result
+     */
+    suspend operator fun invoke(userId: UserId): Flow<GetUserInfoResult>
+}
+
+internal class ObserveUserInfoUseCaseImpl(
+    private val userRepository: UserRepository,
+    private val teamRepository: TeamRepository
+) : ObserveUserInfoUseCase {
+
+    override suspend fun invoke(userId: UserId): Flow<GetUserInfoResult> {
+        return observeOtherUser(userId)
+            .foldEither({ GetUserInfoResult.Failure }) { otherUser ->
+                getOtherUserTeam(otherUser).fold(
+                    { GetUserInfoResult.Failure },
+                    { team -> GetUserInfoResult.Success(otherUser, team) })
+            }
+    }
+
+    private suspend fun observeOtherUser(userId: UserId): Flow<Either<CoreFailure, OtherUser>> {
+        return userRepository.getKnownUser(userId)
+            .wrapStorageRequest()
+            .mapLeft { storageFailure ->
+                if (storageFailure is StorageFailure.DataNotFound) {
+                    userRepository.fetchUsersByIds(setOf(userId))
+                        .fold({ it }) { storageFailure }
+                } else {
+                    storageFailure
+                }
+            }
+            .filter { either ->
+                // We don't want to pass DataNotFound Error forward, cause in that case we'll fetch data from API
+                either.fold({ it !is StorageFailure.DataNotFound }) { true }
```

Comment:
This shouldn't ever happen, because `fetchUsersByIds` isn't supposed to emit `DataNotFound`, only `getKnownUser` can do so, but it's already filtered out in `mapLeft`, so maybe even this filter isn't needed? 😄

Reply:
not really `StorageFailure.DataNotFound` can be emitted in `mapLeft` when `userRepository.fetchUsersByIds` is succeed.

Reply:
I mean that `fetchUsersByIds` currently doesn't emit `DataNotFound` because currently it executes `upsertTeamMembers` and `upsertUsers` - both return `Unit` and `DataNotFound` is passed when something returns `null` 🙂 
I'm just worried because this filter won't pass the `DataNotFound` so if in the future the `fetchUsersByIds` code changes and starts emitting this error then we won't get any value and be stuck on loading for instance 🤔